odoo/modules/packages.py

```diff
@@ -0,0 +1,148 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+""" Modules dependency graph. """
+
+import logging
+# Deprecated since version 3.9:
+# https://docs.python.org/3.9/library/typing.html#typing.Iterable
+from typing import List, Dict, Iterable, Iterator
+
+import odoo
+import odoo.tools as tools
+
+_logger = logging.getLogger(__name__)
+
+
+class Package:
+    """
+    Python package for the Odoo module with the same name
+    """
+    def __init__(self, name: str) -> None:
+        # manifest data
+        self.name = name
+        self.manifest = manifest = odoo.modules.module.get_manifest(name) or {}
+
+        # ir_module_module data             # column_name
+        self.id = None                      # id
+        self.state = 'uninstalled'          # state
+        self.dbdemo = False                 # demo
+        self.installed_version = None       # latest_version
+
+        # info for upgrade
+        self.upgrade_end = False            # if the package is need to upgrade end-xxx
+        self.old_version = None             # the version when added to packages
+
+        # dependencies
+        self.depends = set(manifest.get('depends', ()))
+        self.dependencies: List[Package] = []
```

Comment:
It's actually more subtle :smile: The difference actually is:
```py
self.depends: Set[str]
self.dependencies: List[Package]
```
The fact of using a `set` for `self.depends` makes method `_remove()` faster, since the modules that depend on a given module can be found in O(n).